include/maliput/math/kd_tree.h

```diff
@@ -264,5 +344,148 @@ class KDTree {
   std::deque<Node> nodes_;
 };
 
+/// N-Dimension KDTree.
+/// See KDTreeBase for details.
+template <typename Coordinate, ::std::size_t Dimension,
+          typename Region = BoundingRegion<Coordinate>,
+          typename Distance = details::SquaredDistance<Coordinate, Dimension>,
+          typename NodeCmp = details::NodeCmp<Dimension>>
+class KDTree
+    : public KDTreeBase<KDTree<Coordinate, Dimension, Region, Distance, NodeCmp>, Coordinate, Dimension, Region, Distance, NodeCmp> {
+ public:
+  template <typename Iterator>
+  KDTree(Iterator begin, Iterator end)
+      : KDTreeBase<KDTree<Coordinate, Dimension, Region, Distance, NodeCmp>, Coordinate, Dimension, Region, Distance, NodeCmp>(begin,
+                                                                                                               end) {}
+
+  /// Constructs a KDTreeBase taking a vector of points.
+  ///
+  /// @param points Vector of points
+  /// @tparam Collection type of the collection.
+  /// @throws maliput::common::assertion_error When the range is empty.
+  template <typename Collection>
+  KDTree(Collection&& points)
+      : KDTreeBase<KDTree<Coordinate, Dimension, Region, Distance, NodeCmp>, Coordinate, Dimension, Region, Distance, NodeCmp>(points) {
+  }
+};
+
+/// 3-Dimensional KDTree.
+/// In addition it provides a RangeSearch method for range queries in the 3D-space.
+///  - For using the RangeSearch query some extra initialization is needed.
+///    - InitializeRegions(): Regions should be initialized before starting to use the RangeSearch query.
+///                           This isn't automatically done by the constructor because it is expensive. (Adds ~20% time to construction time)
+/// @code {.cpp}
+///  KDTree<Vector3> tree{points};
+///  tree.InitializeRegions(); // Initialize regions for the RangeSearch query.
+///  tree.RangeSearch(region_1);
+///  tree.RangeSearch(region_2);
+///  ...
+///  tree.Nearest(point_1); // NearestNeighbour (NN).
+/// @endcode
```

Comment:
I don't particularly like having lazy initialization in a way that the user have to run an extra initialization method if he decided to use range query searches.
However, this initialization adds a 20% time to the constructor time, and if the user weren't up to using the range search queries then they would be "wasting" time. 

There are alternatives for sure:
 - Initialize it always automatically. (Like this one, more simple to use.)
 - Initializing this by default and overloading constructors for adding a flag to disable this. So as to provide a way to disable this.